File: src/test/java/bigdecimal/ClientInvoiceAllocationCalculatorTest.java
package bigdecimal;

import org.junit.Before;
import org.junit.Test;

import java.math.BigDecimal;

import static bigdecimal.ClientInvoiceAllocationCalculator.calculateClientInvoiceAllocationErdAmount;
import static org.junit.Assert.assertTrue;

public class ClientInvoiceAllocationCalculatorTest {

    private ClientInvoiceAllocation cia;

    @Before
    public void setUp() throws Exception {
        cia = new ClientInvoiceAllocation();
        cia.setVatAmount(new BigDecimal("20"));
        cia.setInvoiceRate(new BigDecimal("2.2"));
        cia.setVatOutStandingAmount(new BigDecimal("200"));
        cia.setExchangeRate(new BigDecimal("1.6666"));
    }

    @Test
    public void testCalculateClientInvoiceAllocationErdAmount1() throws Exception {
        calculateClientInvoiceAllocationErdAmount(cia);
        assertTrue(cia.getErdAmount().compareTo(new BigDecimal("2.909571")) == 0);
    }

    @Test
    public void testCalculateClientInvoiceAllocationErdAmount2() throws Exception {
        cia.setVatAmount(new BigDecimal("2000"));
        calculateClientInvoiceAllocationErdAmount(cia);
        assertTrue(cia.getErdAmount().compareTo(BigDecimal.ZERO) == 0);
    }
}

File: src/test/scala/bigdecimal/ClientInvoiceAllocationCalculatorTest.scala
package bigdecimal

import org.scalatest.{BeforeAndAfter, FunSuite}
import java.math.MathContext.DECIMAL32
import bigdecimal.ClientInvoiceAllocationCalculator._

class ClientInvoiceAllocationCalculatorTest extends FunSuite with BeforeAndAfter {
  var cia: ClientInvoiceAllocation = _

  before {
    cia = new ClientInvoiceAllocation()
    cia.setVatAmount(new java.math.BigDecimal("20", DECIMAL32))
    cia.setInvoiceRate(new java.math.BigDecimal("2.2", DECIMAL32))
    cia.setVatOutStandingAmount(new java.math.BigDecimal("200", DECIMAL32))
    cia.setExchangeRate(new java.math.BigDecimal("1.6666", DECIMAL32))
  }

  test("java CalculateClientInvoiceAllocationErdAmount") {
    calculateClientInvoiceAllocationErdAmount(cia)
    assert(cia.getErdAmount === new java.math.BigDecimal("2.909571"))
  }

  test("java CalculateClientInvoiceAllocationErdAmount VAT same") {
    cia.setVatAmount(new java.math.BigDecimal("200", DECIMAL32))
    calculateClientInvoiceAllocationErdAmount(cia)
    assert(cia.getErdAmount === new java.math.BigDecimal("29.09571"))
  }

  test("java CalculateClientInvoiceAllocationErdAmount VAT larger") {
    cia.setVatAmount(new java.math.BigDecimal("2000", DECIMAL32))
    calculateClientInvoiceAllocationErdAmount(cia)
    assert(cia.getErdAmount === java.math.BigDecimal.ZERO)
  }
}
